data/modules/TradeShips/Events.lua

```diff
@@ -0,0 +1,325 @@
+	if Core.ships[ship] == nil then return end
+	local trader = Core.ships[ship]
+	trader.ts_error = "HIT"
+
+	trader['chance'] = trader.chance or 0
+	trader['chance'] = trader.chance + 0.1
+
+	-- don't spam actions
+	if trader.last_flee and Game.time - trader.last_flee < Engine.rand:Integer(5, 7) then return end
+
+	-- if outbound jump now
+	if trader.status == 'outbound' then
+		if Trader.getSystemAndJump(ship) == 'OK' then
+			return
+		end
+	end
+
+	trader['status'] = 'fleeing'
+	trader['attacker'] = attacker
+
+	-- update last_flee
+	trader['last_flee'] = Game.time
+
+	-- if distance to starport is far attempt to hyperspace
+	if trader.no_jump ~= true then
+		local starports = Core.params.local_routes
+		if #starports == 0 then
+			trader['no_jump'] = true -- it already tried in onEnterSystem
+		elseif trader.starport and Engine.rand:Number(1) < trader.chance then
+			local distance = ship:DistanceTo(trader.starport)
+			if distance > Core.AU * (2 - trader.chance) then
+				if Trader.getSystemAndJump(ship) then
+					return
+				else
+					trader['no_jump'] = true
+					trader['chance'] = trader.chance + 0.3
+				end
+			end
+		end
+	end
+
+	-- maybe jettison a bit of cargo
+	if Engine.rand:Number(1) < trader.chance then
+		local cargo_type = nil
+		local max_cap = ShipDef[ship.shipId].capacity
+		for k, v in pairs(ship:GetCargo()) do
+			if v > 1 and Engine.rand:Number(1) < v / max_cap then
+				cargo_type = k
+				break
+			end
+		end
+		if cargo_type and ship:Jettison(cargo_type) then
+			Comms.ImportantMessage(attacker.label..', take this and leave us be, you filthy pirate!', ship.label)
+			trader['chance'] = trader.chance - 0.1
+		end
+	end
+end
+Event.Register("onShipHit", onShipHit)
+
+local onShipCollided = function (ship, other)
+	if Core.ships[ship] == nil then return end
+	if other:isa('CargoBody') then return end
+	Core.ships[ship].ts_error = "collided"
+
+	if other:isa('Ship') and other:IsPlayer() then
+		onShipHit(ship, other)
+		return
+	end
+end
+Event.Register("onShipCollided", onShipCollided)
+
+local onShipDestroyed = function (ship, attacker)
+	if attacker == nil then return end-- XX
+	if Core.ships[ship] ~= nil then
+		local trader = Core.ships[ship]
+		Core.log:add(ship, 'Destroyed by '..attacker.label..', status:'..trader.status..' starport:'..(trader.starport and trader.starport.label or 'N/A'))
+		Core.ships[ship] = nil
+		-- XXX consider spawning some CargoBodies if killed by a ship
+	else
+		for t_ship, trader in pairs(Core.ships) do
+			if trader.attacker and trader.attacker == ship then
+				trader['attacker'] = nil
+				if trader.status == 'fleeing' then
+					-- had not reached starport yet
+					trader['status'] = 'inbound'
+					trader.ts_error = "attack!"
+				elseif trader.status == 'cowering' then
+					-- already reached starport and docked
+					-- we need to do everything that the ship does after docking
+					onShipDocked(t_ship, trader.starport)
+				end
+				return
+			end
+		end
```

Comment:
This seems like an unnecessarily expensive function; I would suggest maintaining a cache of attacking ships to test if we need to iterate all tradeships in a system. Making it a table with weak keys would probably be the best route here.

Reply:
Made a cache as a table, where the key is the object of the attacker, and the value is an array of weak links that were added to the traders.